Add Embedded Workflows content (#1752)
ManageIQ/manageiq-documentation: managing_providers/_topics/embedded_workflows.md

## Embedded Workflows

> *Embedded Workflows* provide the capability to simplify service catalog creation and improve the overall service catalog experience through a step-based automation system. Each step within a workflow runs a container to complete processes. Workflows are imported from GitHub repository content into {{ site.data.product.title_short }} where they are made available for users to run.

Reviewer: > Each step within a workflow runs a container to complete processes.

This isn't exactly true, Task steps will run a container but there are other types of steps.  You could say something like "Workflows can run docker containers to accomplish task the user requires as part of the state machine"

Reviewer: > Workflows are imported from GitHub repository content

Same comment as above re: GitHub

Author: For the first, the sentence is updated to be: Workflows can run docker containers to accomplish a task that the user requires as part of the state machine. 

For the second, this is updated to be just Git repository.

   This page lists, and provides a summary of, the credentials for accessing and using embedded workflows. From this page, you can add, update, and delete credentials.

In addition to the *Embedded Workflows* UI, **Embedded Workflows** can also be used in Service Dialogs for dynamic dialog fields and with Service Catalog Items for provisioning, reconfiguration, and retirement entrypoints.

* Service Dialog

   When creating a service dialog, fields that are dynamic include the option for using **Embedded Workflows** or **Embedded Automate**.

   ![Service Dialog options](../images/embedworkflow_servicedialogdynamic.png)

* Service Catalog Items

   The provisioning, reconfigure, and retirement entry points include the option for using **Embedded Workflows** or **Embedded Automate**.  You can set the point field to the workflow that you want to use for any, or all, of the three service entry points.

   ![Entry Points options](../images/embedworkflow_entrypointsembedded.png)

### Adding an Embedded Workflows Repository

> As a starting point to begin authoring a workflow, you must first add the repository that includes your workflows.

Author: This is updated as requested.

To add a repository, complete the following steps:

1. Click  **Automation** > **Embedded Workflows** > **Repositories** > **Configuration** > **Add New Repository**.

   ![Add a repository](../images/embedworkflow_addnewrepository.png)

2. Enter the repository name in the **Name** field.

3. Add a description for the repository in the **Description** field.

> 4. Enter a Git repository URL for the repository in the **GitHub URL** field. The Git repository URL does not need to be a GitHub repository.

Reviewer: The field is just "URL" not "GitHub URL" https://github.com/ManageIQ/manageiq-documentation/pull/1752/files#diff-e473dd8e1cafe1f15527bc795eea54496c4e09ff40416a770384f11edb7a65f5

```suggestion
4. Enter a Git repository URL for the repository in the **URL** field. The Git repository URL does not need to be a GitHub repository.
```

Author: This is updated as requested.

6. Select the appropriate SCM (Source Control Management) credentials from the **SCM Credentials** dropdown menu. These are the credentials for how you access your specified Git repository, and are used in different ways within embedded workflows.

>    If you are adding a repository that is not publicly available, add an SCM (source control) credential for accessing the repository.

Reviewer: These lines feel backwards, it sounds like you are saying "always select an SCM credential, then if you are adding one that is not publicly available select an scm credential".  Maybe `6. If you are adding a repository that requires authentication then add an SCM credential and pick that for this repository`

Author: This step is updated to be:
If you are adding a repository that requires authentication then add an SCM (Source Control Management) credential for accessing the repository. Select that credential for this repository from the **SCM Credentials** dropdown menu. These credentials are used in different ways within embedded workflows.

If any workflows are found, click the corresponding number to view the list of workflows.

   ![Repository workflows](../images/embedworkflow_workflownumbers.png)

Alternatively, click **Automation** > **Embedded Workflows**. Then, click **Workflows** to see the list of workflows across all repositories.

### Refreshing Repositories

{{ site.data.product.title_short }} allows you to refresh a targeted repository or all repositories in your inventory to ensure that your workflows are current.

Refresh a targeted repository:

1.  Browse to the menu: **Automation** > **Embedded Workflows** > **Repositories**.

2.  Click a repository.

3.  Click **Configuration**, then ![Refresh this Repository](../images/2003.png) (**Refresh this Repository**).

Alternately, you can refresh some or all repositories from the list view:

1.  Browse to the menu: **Automation** > **Embedded Workflows** > **Repositories**.

2.  Check those repositories to refresh. Click **Check All** to select all repositories.

3.  Click **Configuration**, then ![Refresh Selected Workflow Repositories](../images/2003.png)(**Refresh Selected Workflow Repositories**).

### Adding Credentials

> Credentials are used by {{ site.data.product.title_short }} for any authentication that is required when running embedded workflows, and when importing workflows from a version control system.

Reviewer: ```suggestion
Credentials are used by {{ site.data.product.title_short }} for any authentication that is required when running embedded workflows or when importing workflows from a version control system.
```

Author: This is updated as suggested

> {{ site.data.product.title_short }} can store credentials that are used by embedded workflows. Credentials that are saved in {{ site.data.product.title_short }} are matched and executed with a workflow when it runs.

Reviewer: ```suggestion
{{ site.data.product.title_short }} can store credentials that are used by embedded workflows. Credentials that are saved in {{ site.data.product.title_short }} are mapped to values in the `Credentials` field of your Workflow and are made available to the container with a workflow when it runs.
```

Author: I have updated this sentence.

If you need to edit a credential, select the credential from the list of credentials. Then, click **Configuration** > **Edit this Credential**. After you complete your edits, click **Save**.

If you need to delete a credential, select the credential from the list of credentials. Then, click **Configuration** > **Remove selected Credentials from Inventory**.

#### Credential Types

Each credential type that is used by {{ site.data.product.title_short }} for embedded workflows is detailed in the following sections.

##### SCM

SCM (source control Management) credentials are used with Projects to clone and update local source code Git repositories. Source Control credentials have several attributes that can be configured:

- **Username**: The username to use in conjunction with the source control system.

- **Password**: The password to use in conjunction with the source control system.

- **Private key passphrase**: If the SSH private key used is protected by a passphrase, you can configure a key passphrase for the private key.

- **Private Key**: Copy or drag-and-drop the actual SSH Private Key to be used to authenticate the user to the source control system by using SSH.

> ### Authoring a workflow

Reviewer: It looks like we lost the "##### Workflows` credential type here, the only type I see is `SCM`

Author: All other credential types were removed based on previous feedback. Should any be added or the section heading/sentence removed and rewritten to merge into adding credentials section.

You can configure workflows for use in the following areas during service catalog creation:

- **Dynamic dialog fields** - You (or your users) can attach workflows to a field so that the field becomes populated or refreshed with the results from the running of the workflow.
- **Service Entry Points** - You (or your users) can set and use workflows for provisioning, reconfiguring, and retiring services.

> The Automate model provides flexibility to not only change parts of the provisioning process, but also to allow you to automate other operational tasks. You can create and use your own embedded workflows as needed to automate your operational tasks.

Reviewer: 2x `automate other operational tasks` feels repetitive repetitive

Author: I have rewritten this content to remove  the duplication: 
You can create and use embedded workflows as needed to not only change parts of the provisioning process, but also to automate other operational tasks.

> Workflows must be authored in Amazon State Languages (ASL) format. As part of authoring a workflow, you (or your users) must build and support the containers that are run during each step of an embedded workflow.

Reviewer: `you (or your users) must build and support the containers that are run during each step of an embedded workflow.`

Maybe put a positive spin on this?  `you (or your users) can build container images that are able to perform any tasks required in any language you like and use these images during Task states in your workflows.`

Our goal was to allow people to do this, not require them to do this after all :wink:

Author: I have made this update as requested.

> 1. Define the code for the workflow. If your workflow requires the use of any credentials or parameters to be specified, ensure that they are passed in the code.

Reviewer: `If your workflow requires the use of any credentials or parameters to be specified, ensure that they are passed in the code.`

This feels like an odd place for this, we haven't even covered the ASL guide but we're talking about parameters and credentials.  Maybe move it down from this line to one below?

Author: I have moved the sentences as suggested.

   The workflow code must be in the Amazon States Language (ASL) format and follow its supported specifications. For more information about Amazon States Language and its specification, see [Amazon States Language Guide](https://states-language.net/).

>    Within the workflow code, you need to specify the states that your workflow requires, including any next steps. For `Task` type steps in the workflow, a docker container is called. The container defines what happens for that Task state. For example, a docker container can run to clone a template.

Reviewer: ```suggestion
   Within the workflow code, you need to specify the states that your workflow requires, including any next steps. For `Task` type steps in the workflow, a docker container is called. The container defines what happens for that Task state. For example, a docker container can run to clone a template.  If your states require parameters or credentials you can specify them in your State definitions.
```

Author: I have made this requested change.

2. Build the docker containers that are required for the workflow.

<!--
##### Add docker images to a podified deployment and appliances (workflow registries)
Include:
* Building docker containers (reference Docker documentation)
* Push to a registry
* Add the registry to the appliance/kubernetes system.

> -->

Reviewer: ```
Once you have the code for your task resource written you need to bundle it into a docker image.  You can do this by creating a standard Dockerfile (https://docs.docker.com/engine/reference/builder/) and building the image (https://docs.docker.com/engine/reference/commandline/build/).  Then you can push the image to a registry (https://docs.docker.com/engine/reference/commandline/push/) which makes them available to be used by {{ site.data.product.title_short }}.

Once you have pushed your images to an image registry you can add the registry to {{ site.data.product.title_short }}.  On appliances `podman` is used to execute the containers so you would `podman login` (https://docs.podman.io/en/stable/markdown/podman-login.1.html) as the `manageiq` user.
# su manageiq
$ podman login docker.io
Username: 
Password:
Login Succeeded!

It is recommended that you use a docker.io access token (https://docs.docker.com/security/for-developers/access-tokens/) so that the token doesn't expire.
```

TODO kubernetes pull secrets for podified

Author: I have added this content.

   4. Use the drop-down fields to map one or more credentials for the workflow. Verify that the table is updated with the correct information.

   5. Click **Save**.

### Creating a Service Dialog using an Embedded Workflow

Fields that are *Dynamic* offer the choice of *Embedded Workflows* in addition to the original *Embedded Automate*.

Dialogs are the part of {{ site.data.product.title_short }} that you interface with when you order services.

1. To open a service dialog, click **Automation** > **Embedded Automate** > **Customization**.
2. From the *Service Dialog* section, review the list of created dialogs. Click one of the dialogs to view the details.
3. Click **Edit the dialog**. You can edit the *Name*, *Provider*, and *Template*.

   When you select to edit any of the dropdown settings, you can see a *dynamic* option. Toggle to **On** to make sure it is enabled.

4. Click the **Options** tab. You can now select between *Embedded Automate* and *Embedded Workflows* in the Automation field.

   ![Dialog Dynamic](../images/embedworkflow_servicedialogdynamic.png)

>    If you choose **Embedded Workflows**, you can select the workflow to run. Complete the fields, such as *Template* (using drag and drop functionality) and click **Save** when you complete the form. The workflow now runs and refreshes the fields with data from the workflow. In this way fields are chained together. When you *order* a service catalog, the workflow runs in the background to refresh the template.

Reviewer: > Complete the fields, such as *Template* (using drag and drop functionality) and click **Save** when you complete the form.

Don't think we should call out Template specifically since that was just an example: `Complete the rest of the fields and click **Save** when you complete the form.`

> The workflow now runs and refreshes the fields with data from the workflow.

The workflow won't run when you click save you have to order the service first, you cover this later on so maybe just drop this sentence?

> In this way fields are chained together.

This feels like part of a though, maybe `You can chain dialog fields together by using **Fields to refresh**, this allows one dialog field to kick off the run of another if there are inter-dependencies`

> When you *order* a service catalog, the workflow runs in the background to refresh the template.

Again the template was just an example, `When you *order* a service catalog any dynamic dialogs that are backed by embedded workflows will be run in the background to populate the dialogs.`

Author: I have updated this content as suggested.

### Creating Service Catalog Items using an Embedded Workflow

You can create a generic service catalog item that uses an embedded workflow. To use previously set up workflows and to view a service catalog item, you need to *order* the item. Complete the following steps to order a service catalog item that uses embedded workflows:

1. Click **Services** > **Catalogs**.
2. The *Service Catalogs* section opens and you see the existing Service Catalogs. If you want to order one, click it and then **Order**.

   ![Service Dialog Dynamic](../images/embedworkflow_orderservicecatalog.png)

   Complete the fields for your specified dialog. Then, click **Submit**.

3. If you want to know more about the item first or to edit it, click the **Catalogs** section, and then click that item that is created. Click **Configuration** > **Edit this item**.

   ![Entry Points Dynamic](../images/embedworkflow_entrypointsembedded.png)

4. When you complete all your edits, click **Save**.

> 5. Confirm that you can order the service catalog item and that it runs the workflows to drive the dynamic dialog dropdown.

Reviewer: :+1: good idea, here is a screenshot of what that looks lie
![image](https://github.com/ManageIQ/manageiq-documentation/assets/12851112/2870a4ca-cf6f-4dde-9001-6ed5eca4bc82)

> 5. Confirm that you can order the service catalog item and that it runs the workflows to drive the dynamic dialog dropdown.

Author: I have added this statement and screenshot.

## Embedded Workflows

> {: #embedded_workflows}

Reviewer 2: @rwellon What is this symbol?  When I view it rendered, it shows up.

![manageiq-documentation_managing_providers__topics_embedded_workflows_md_at_10f4ce35938dffb91aef18139c2dcca861de5869_·_ManageIQ_manageiq-documentation](https://github.com/ManageIQ/manageiq-documentation/assets/52120/d6bf57cc-507a-4df4-94e7-4341cba068fd)

Author: In the downstream IA docs we use the anchors for adding links to the headings within topics. I notice in ManageIQ docs this is not used. I removed this anchor.

*Embedded Workflows* are an alternative to the *Embedded Automate* capability for authoring and running automation. With *Embedded Workflows*, you can continue to use the *Embedded Automate* feature for service catalog creation and management.

### Enabling Embedded Workflows

The Embedded Workflows functionality is not enabled and available by default. A user with administrator rights must enable the functionality by turning on a prototype flag. If you are an administrative user, you can complete the following steps to enable the functionality:

1. From the {{site.data.product.title_short }} UI, click **Settings** > **Applications Settings**. Then, click the **Advanced** tab.
2. On the *Advanced* tab, search for `ems_workflows` within the *Prototype* section. For the entry, set the `enabled` setting to `true`.
3. Click **Save**.
4. Restart the application.

> When the application is restarted, the *Embedded Workflows* functionality is enabled. <!-- This functionality reuses existing API. -->

Reviewer: Embedded Workflows are enabled by default now (as of quinteros) so the user doesn't have to change any settings for this to be available.

https://github.com/ManageIQ/manageiq-providers-workflows/pull/52

Author: I have removed this section, since it is not needed anymore.

> When the application is restarted, the *Embedded Workflows* functionality is enabled. <!-- This functionality reuses existing API. -->

Reviewer 2: This is no longer true - we can remove this statement now as it is on by default.

Author: I have removed this section as requested.

### Embedded Workflows UI additions

> When the *Embedded Workflows* functionality is enabled, new **Automation** > **Embedded Workflows** pages become available for use in the UI. In addition to the *Embedded Workflows* UI, you can now select **Embedded Workflows** as a dropdown option along with **Embedded Automate** for various fields on other UI pages.

Reviewer: Not necessary to say "when the functionality is enabled" per above

Author: I have removed/updated this statement.

> When the *Embedded Workflows* functionality is enabled, new **Automation** > **Embedded Workflows** pages become available for use in the UI. In addition to the *Embedded Workflows* UI, you can now select **Embedded Workflows** as a dropdown option along with **Embedded Automate** for various fields on other UI pages.

Reviewer 2: And since it's always enabled we can remove the "When the *Embedded Workflows* functionality is enabled,"

#### Embedded Workflow UI

> The new *Embedded Workflows* functionality adds the following pages to the UI, which you can use to access and work with workflows. To view these pages, go to **Automation** > **Embedded Workflows** and select the page.

Reviewer: I don't think we need "The new ..." since it won't be "new" for very long but we are guaranteed to forget to update the docs :smile:

> The new *Embedded Workflows* functionality adds the following pages to the UI, which you can use to access and work with workflows. To view these pages, go to **Automation** > **Embedded Workflows** and select the page.

Reviewer 2: Since documentation persists forever, I think we should remove the "new" phrasing.  In the next release it won't be new anymore and we'll have to go back and remember to remove it.

Author: I have removed/updated this statement.

> ![Embedded Workflow UI](../images/embedworkflow_ui.png){: caption="Embedded Workflow UI" caption-side="bottom"}

Reviewer 2: This caption notation doesn't work in Markdown (i.e. it's rendered as is)

Reviewer 2: FWIW, alt text is handled by the text between the `[]` in the Markdown image link

Author: This caption style is used by the downstream markdown - I have removed it from this file.

- **Workflows**

   This UI page includes a summary of the available workflows. You can use different views to review the workflows:

   - Text view, which is the raw representation (JSON or YAML) of the Amazon State Languages (ASL) code.

>    - Graph view, which shows an SVG visual representation of the workflow structure.

Reviewer 2: I don't think it's an SVG, so I'm thinking just remove that detail.

```suggestion
   - Graph view, which shows a visual representation of the workflow structure.
```

- **Repositories**

   This page lists, and provides a summary of, the added embedded workflow repositories. From this page, you can add, update, and remove repositories.
   When a repository is added and synchronized, any workflows that are found in the repository can be viewed on the **Workflows** UI page.

- **Credentials**

>    Credentials are used by {{ site.data.product.title_short }} for authentication when running embedded workflows against machines, synchronizing with inventory sources, and importing project content from a version control system.

Reviewer: > for authentication when running embedded workflows against machines,

This feels overly simplified (maybe just copy&paste from the ansible side), they are for any authentication needs that a workflow has.

> synchronizing with inventory sources

I think this was also copied form ansible and isn't applicable at all here.

> and importing project content from a version control system.

`and importing workflows from a version control system`

   This page lists, and provides a summary of, the credentials for accessing and using embedded workflows. From this page, you can add, update, and delete credentials.

>    These credentials are tied to the provider instance, and are used by the workflow when it runs. The credential types are defined by the provider.

Author: I have removed this statement: `These credentials are tied to the provider instance, and are used by the workflow when it runs. The credential types are defined by the provider.`

> #### UI Field Updates

Reviewer: I don't think we should say "UI Field Updates" because this reads more like a changelog (aka here is what we added) and less like long-term documentation.

Author: I have removed this header and updated the section.

{: #field_updates}

> Existing integration points are updated for the service catalog creation and management process to support choosing to use **Embedded Workflows** or **Embedded Automate**. This choice can be set when working with service catalog items and service dialogs.

Reviewer: ```suggestion
**Embedded Workflows** can be used in Service Dialogs for dynamic dialog fields and Service Catalog Items for provisioning, reconfiguration, and retirement entrypoints
```

Author: I have made this change.

* Service Dialog

   When creating a service dialog, fields that are dynamic now include the option for using **Embedded Workflows** or **Embedded Automate**.

   ![Service Dialog options](../images/embedworkflow_servicedialogdynamic.png){: caption="Service Dialog options" caption-side="bottom"}

* Service Catalog Items

>    The provisioning, configuring, or retiring entry points now include the option for using **Embedded Workflows** or **Embedded Automate**.

Reviewer: ```suggestion
   The provisioning, reconfigure, and retirement entry points now include the option for using **Embedded Workflows** or **Embedded Automate**.
```

Reviewer 2: "now include" is similar wording to "new" in that it's going to get stale...wondering if just taking out the "now" is sufficient.  (Similarly for the other sections here).

Author: I have made this change.

   ![Entry Points options](../images/embedworkflow_entrypointsembedded.png){: caption="Entry Points options" caption-side="bottom"}

### Adding an Embedded Workflows Repository

As a starting point to begin authoring a workflow, you must first add the repository that includes your workflows.

To add a repository, complete the following steps:

1. Click  **Automation** > **Embedded Workflows** > **Repositories** > **Configuration** > **Add New Repository**.

   ![Add a repository](../images/embedworkflow_addnewrepository.png){: caption="Add a repository" caption-side="bottom"}

> 2. Enter *Name*, *Description*, and a *GitHub URL*.

Reviewer: You can enter any git repository URL not just a GitHub one

Author: I have updated the steps for this procedure to indicate it is not specifically a GitHub URL.

> Enter a Git repository URL for the repository in the **GitHub URL** field. The Git repository does not need to be a GitHub repository.

3. Provide a Repository Name in the **Name** field.

4. Add a description for the repository in the **Description** field.

> 5. Add a **GitHub URL** for the repository.

Reviewer: Are we adding the URL twice?

Author: I have updated the procedure to remove the duplicated step.

> 6. Select the appropriate SCM (Source Control) credentials from the **SCM Credentials** dropdown menu. These are the credentials for how you access GitHub, and are used in different ways within embedded workflows.

Reviewer: > These are the credentials for how you access GitHub

These credentials are not just for GitHub they are for any repository that you have, e.g. Gitlab or Bitbucket are other publicly hosted options, and anyone can run a private git server.

7. Provide a branch name in the **SCM Branch** field for the branch that you want to pull.

> 8. Click **Save**. Upon saving, the repository is imported and scanned for any included workflows.

Reviewer: We should probably mention that "any included workflows" means `.asl` files

Author: I have updated this text to be: 

> lick **Save**. Upon saving, the repository is imported and scanned for any included workflows (any Amazon States Language `.asl` files).

   ![Repository details](../images/embedworkflow_addnewrepository_details.png){: caption="Repository details" caption-side="bottom"}

9. Confirm that the synchronization with your repository is successful by viewing the list of workflows that are detected for the repository.

> **Note:** If you are adding a repository that is not publicly available, add an SCM (source control) credential for accessing the repository. This credential is separate from the workflow credentials that you configure in the procedure for adding the repository. Source Control credentials have several attributes that can be configured:

Reviewer: > This credential is separate from the workflow credentials that you configure in the procedure for adding the repository.

Hm?  When adding a repository you would be optionally adding an SCM credential not a Workflow credential

Reviewer: I feel like this should be merged into line 84 above https://github.com/ManageIQ/manageiq-documentation/pull/1752/files#diff-a944ad335565b3b6f0c1776d24ce585d3b42531e2d45bc20a5eaab1950a871e4R84

Author: I have merged this content as suggested.

   - **Username**: The username to use in conjunction with the source control system.
   - **Password**: The password to use in conjunction with the source control system.
   - **Private key passphrase**: If the SSH private key used is protected by a passphrase, you can configure a key passphrase for the private key.
   - **Private Key**: Copy or drag and drop the actual SSH Private Key to be used to authenticate the user to the source control system by using SSH.

### Viewing the List of Workflows in a Repository

> When your repository is added, the repository shows in the *Repositories* list. You can now click the entry for the repository to open it so that you can scan for existing workflows.

Reviewer: > so that you can scan for existing workflows.

This sounds like clicking on the repository is initiating the scan (maybe just overloaded terms).  How about "so that you can view the imported workflows"

Author: I have updated this sentence to be: You can now click the entry for the repository to open it so that you can view the imported workflows.

> If multiple workflows are found, click the corresponding number to view the list of workflows, **3** in this case.

Reviewer: > If multiple workflows are found, click the corresponding number

This will be true even if only 1 workflow is found, `if multiple` doesn't make sense here.

> , **3** in this case.

In what case?  I don't think we've given them the ManageIQ/workflows-examples repo as a demo here we can probably just leave this part off.

   ![Repository workflows](../images/embedworkflow_workflownumbers.png){: caption="Repository workflows" caption-side="bottom"}

> Alternatively, click **Automation** > **Embedded Workflows**. Then, click **Workflows** to see the workflows list.

Reviewer: > to see the workflows list.

Probably want to mention that `**Automation** > **Embedded Workflows**` will be the list of workflows across all Repositories

Author: I have updated this step to be:

> Alternatively, click **Automation** > **Embedded Workflows**. Then, click **Workflows** to see the list of workflows across all repositories.

### Refreshing Repositories

{{ site.data.product.title_short }} allows you to refresh a targeted repository or all repositories in your inventory to ensure that your workflows are current.

Refresh a targeted repository:

1.  Browse to the menu: **Automation** > **Embedded Workflows** > **Repositories**.

2.  Click a repository.

3.  Click **Configuration**, then ![Refresh this Repository](../images/2003.png) (**Refresh this Repository**).

Alternately, you can refresh some or all repositories from the list view:

1.  Browse to the menu: **Automation** > **Embedded Workflows** > **Repositories**.

2.  Check those repositories to refresh. Click **Check All** to select all repositories.

3.  Click **Configuration**, then ![Refresh Selected Workflow Repositories](../images/2003.png)(**Refresh Selected Workflow Repositories**).

### Adding Credentials

> Credentials are used by {{ site.data.product.title_short }} for authentication when running embedded workflows against machines, synchronizing with inventory sources, and importing project content from a version control system.

Reviewer: Is this a duplicate of https://github.com/ManageIQ/manageiq-documentation/pull/1752/files#diff-a944ad335565b3b6f0c1776d24ce585d3b42531e2d45bc20a5eaab1950a871e4R43 ?

If we need to have both then the same comment applies

Author: I have made the same change in this section.

{{ site.data.product.title_short }} can store credentials that are used by embedded workflows. Credentials saved in {{ site.data.product.title_short }} are matched and executed with a workflow when it runs.

1.  Browse to the menu: **Automation** > **Embedded Workflows** > **Credentials**.

2.  Click **Configuration**, then ![Add New Credential](../images/1862.png) (**Add New Credential**).

3.  Provide a **Name** for the credential.

4.  Select the **Credential Type**. More fields appear depending on the type chosen.

5.  Click **Add**.

> #### Credential Types

Reviewer 2: Not all of the credentials below are used by Embedded Workflows.  All of them are Ansible however, so I'd be good if we kept that content for Embedded Ansible.

Workflows only has "SCM" and "Workflow" types.

Author: I have removed the types other than SCM (workflow needs to be added) as requested.

- **Privilege Escalation**: Specifies the type of escalation privilege to assign to specific users. Options include **sudo**, **su**, **pbrun**, **pfexec**.

- **Privilege Escalation Username**: Enter the username to use with escalation privileges on the remote system.

- **Privilege Escalation Password**: Enter the actual password to be used to authenticate the user through the selected privilege escalation type on the remote system.

##### Network

Network credentials are used for networking modules to connect to and manage networking devices. Network credentials have several attributes that can be configured:

- **Username**: The username to use in conjunction with the network device.

- **Password**: The password to use in conjunction with the network device.

- **Authorize**: Select this from the Options field to add an Authorize password that signs the RSA key with a password.

- **Authorize password**: If **Authorize** is checked, enter a password in the **Authorize Password** field.

- **SSH Key**: Copy or drag-and-drop the actual SSH Private Key to be used to authenticate the user to the network by using SSH.

> - **Private key passphrase**: The actual passphrase for the private key to be used to authenticate the user to the network by using SSH.

Author: I have removed these types as requested.

##### SCM

> SCM (source control) credentials are used with Projects to clone and update local source code repositories from a remote revision control system such as Git, Subversion, or Mercurial. Source Control credentials have several attributes that can be configured:

Reviewer: @Fryguy do we actually support svn/mercurial?  I think we only have `GitRepository` at least in the modeling

Author: I have updated this to be: SCM (source control Management) credentials are used with Projects to clone and update local source code Git repositories.

Selecting this credential type enables synchronization of cloud inventory with Red Hat Virtualization. Red Hat Virtualization credentials have several attributes that can be configured:

- **Username**: The username to use to connect to Red Hat Virtualization.

- **Password**: The password to use to connect to Red Hat Virtualization.

- **Host (Authentication URL)**: The host to be used for authentication.

   **Important**: Enter in **Host** the Red Hat Virtualization provider URL, followed by the path `/ovirt_engine/api`. Example: `https://your.rhv.com/ovirt_engine/api`

##### VMware

Selecting this credential type enables synchronization of inventory with VMware vCenter.

**Important:** If both {{ site.data.product.title_short }} and a VMware provider are located in the same IPv6-only network, use a DNS-resolvable hostname for the VMware provider in the **vCenter Host** field when adding credentials.

VMware credentials have several attributes that can be configured:

- **Username**: The username to use to connect to vCenter.

- **Password**: The password to use to connect to vCenter.

- **vCenter Host**: The vCenter hostname or IP address to connect to.

> **Note:** If the VMware guest tools are not running on the instance, VMware inventory sync might not return an IP address for that instance.

Reviewer: These are all from Ansible and not used by workflows and can be deleted from here

Author: I have removed these types as requested.

- **Dynamic dialog fields** - You (or your users) can attach workflows to a field so that the field becomes populated or refreshed with the results from the running of the workflow.
- **Service Entry Points** - You (or your users) can set and use workflows for provisioning, reconfiguring, and retiring services.

The Automate model provides flexibility to not only change parts of the provisioning process, but also to allow you to automate other operational tasks. You can create and use your own embedded workflows as needed to automate your operational tasks.

Workflows must be authored in Amazon State Languages (ASL) format. As part of authoring a workflow, you (or your users) must build and support the containers that are run during each step of an embedded workflow.

1. Define the code for the workflow. If your workflow requires the use of any credentials or parameters to be specified, ensure that they are passed in the code.

   The workflow code must be in the Amazon States Language (ASL) format and follow its supported specifications. For more information about Amazon States Language and its specification, see [Amazon States Language Guide](https://states-language.net/).

>    Within the workflow code, you need to specify the states that your workflow requires, including any next steps. For each of the steps in the workflow, a docker container is called. The container defines what happens for that state. For example, a docker container can run to clone a template.

Reviewer: > For each of the steps in the workflow, a docker container is called.

A Docker container is only run for `Task` type states, not every state.

Author: I have updated this text from:
> Within the workflow code, you need to specify the states that your workflow requires, including any next steps. For each of the steps in the workflow, a docker container is called. The container defines what happens for that state.

to be: 

> Within the workflow code, you need to specify the states that your workflow requires, including any next steps. For `Task` type steps in the workflow, a docker container is called. The container defines what happens for that Task state.

2. Build the docker containers that are required for the workflow.

#### Example: Provisioning Workflow

```asl
{
  "Comment": "Provision a VMware VM.",
  "StartAt": "CloneTemplate",
  "States": {
    "CloneTemplate": {
      "Type": "Task",

>       "Resource": "docker://docker.io/agrare/clone-template:latest",

Reviewer 2: Let's switch these @agrare specific links to be ManageIQ (here and below).  We will be publishing these to our official ManageIQ account, so let's use manageiq as the docker org, and then all images will be prefix with "workflows-examples-".  For example, let's change this one to

```suggestion
      "Resource": "docker://docker.io/manageiq/workflows-examples-clone-template:latest",
```

Author: I have updated these examples as requested.

        "TASK.$": "$.task"
      }
    },

    "PollTaskComplete": {
      "Type": "Choice",
      "Choices": [
        {
          "Variable": "$.state",
          "StringEquals": "success",
          "Next": "PowerOnVM"
        },
        {
          "Variable": "$.state",
          "StringEquals": "running",
          "Next": "RetryState"
        },
        {
          "Variable": "$.state",
          "StringEquals": "error",
          "Next": "FailState"
        }
      ],
      "Default": "FailState"
    },

    "RetryState": {
      "Type": "Wait",
      "Seconds": 5,
      "Next": "CheckTaskComplete"
    },

    "PowerOnVM": {
      "Type": "Task",
      "Resource": "docker://docker.io/agrare/power-on-vm:latest",
      "Next": "SuccessState",
      "Credentials": {
        "vcenter_user.$": "$.vcenter_user",
        "vcenter_password.$": "$.vcenter_password"
      },
      "Parameters": {
        "VCENTER_HOST.$": "$.vcenter_host",
        "VM.$": "$.vm"
      }
    },

    "FailState": {
      "Type": "Fail",
      "Error": "FailStateError",
      "Cause": "No Matches!"
    },

    "SuccessState": {
      "Type": "Succeed"
    }
  }
}
```

### Viewing workflow details

> When workflows are available, you can view the details for each workflow, such as to verify the payload. To view the details for a workflow:

Reviewer: The "when they are available" feels extraneous here, I wouldn't think I could view details about something unavailable

Maybe,
```suggestion
Once a workflow is imported you can view the details such as to verify the payload. To view the details for a workflow:
```

1. Click **Automation** > **Embedded Workflows**. Then, click **Workflows** to see the workflows list.

2. Click a workflow. The default *Text* view of the workflow opens where you see the code content.

   ![Text view](../images/embedworkflow_textview.png){: caption="Text view" caption-side="bottom"}

   Alternatively, click the **Graph** tab to see a graph of the same data.

   ![Graph view](../images/embedworkflow_graphView.png){: caption="Graph view" caption-side="bottom"}

3. If your workflow has credential fields, then create a workflow credential for each of the fields. Then, map the workflow credential to the workflow.

>    <!-- Add substeps from development - A.G to provide.-->

Reviewer: I think we can reference the Credentials section here saying that you need to create a Workflow Credential, then here in this section we can cover the `Map Credentials` UI

Author: I have updated this section to reference/link to the adding credentials section and added a new step for mapping the credentials. This content will be added later.

### Creating a Service Dialog using an Embedded Workflow

> With Embedded Workflows enabled, fields that are *Dynamic* offer the choice of *Embedded Workflows* in addition to the original *Embedded Automate*.

Reviewer: ```suggestion
Fields that are *Dynamic* offer the choice of *Embedded Workflows* in addition to the original *Embedded Automate*.
```

1. To open a service dialog, click **Automation** > **Embedded Automate** > **Customization**.
2. From the *Service Dialog* section, review the list of created dialogs. Click one of the dialogs to view the details.
3. Click **Edit the dialog**. You can edit the *Name*, *Provider*, and *Template*.

   When you select to edit any of the dropdown settings, you can see a *dynamic* option. Toggle to **On** to make sure it is enabled.

4. Click the **Options** tab. You can now select between *Embedded Automate* and *Embedded Workflows* in the Automation field.

   ![Dialog Dynamic](../images/embedworkflow_servicedialogdynamic.png){: caption="Dialog Dynamic" caption-side="bottom"}

   If you choose **Embedded Workflows**, you can select the workflow to run. Complete the fields, such as *Template* (using drag and drop functionality) and click **Save** when you complete the form. The workflow now runs and refreshes the fields with data from the workflow. In this way fields are chained together. When you *order* a service catalog, the workflow runs in the background to refresh the template.

### Creating Service Catalog Items using an Embedded Workflow

> You can create a generic service catalog item that uses a created dynamic dialog and an embedded workflow. To use previously set up workflows and to view a service catalog item, you need to *order* the item. Complete the following steps to order a service catalog item that uses embedded workflows:

Reviewer: > that uses a created dynamic dialog 

It isn't possible to use a dynamic dialog that isn't created, so I think the `uses a created dynamic dialog` is redundant

Author: I have removed this text. The sentence is now: 

> You can create a generic service catalog item that uses an embedded workflow.

1. Click **Services** > **Catalogs**.
2. The *Service Catalogs* section opens and you see the existing Service Catalogs. If you want to order one, click it and then **Order**.

   ![Service Dialog Dynamic](../images/embedworkflow_orderservicecatalog.png){: caption="Service Dialog Dynamic" caption-side="bottom"}

>    Ensure that the *VM Name*, *Provider* and *Template* fields are correct, and then click **Submit**.

Reviewer: This is very specific to the example dynamic dialog that I've created have and is not generic.  We can probably leave this out or simply say "fill in the fields in your specified dialog then click submit" or something like that

Author: I have updated this sentence as requested. It is now:

> Complete the fields for your specified dialog. Then, click **Submit**.

> 3. If you want to know more about the item first or to edit it, click the **Catalogs** section, and then click that item that is created. Click **Configuration** > **Edit this item**. For instance, one area that you can edit is the three service entry point fields. One field is for *Provisioning*, one for *Reconfiguring* and one for *Retirement*. You can now set these fields to *Embedded Workflows* instead of *Embedded Automate*. You can set the point field to the workflow that you want to use for any, or all, of the three service entry points.

Reviewer: This feels a little redundant (talking about the three entrypoints being able to select Embedded Workflows), :point_up: https://github.com/ManageIQ/manageiq-documentation/pull/1752/files#diff-a944ad335565b3b6f0c1776d24ce585d3b42531e2d45bc20a5eaab1950a871e4R60-R64

Author: I have removed this content: 

> For instance, one area that you can edit is the three service entry point fields. One field is for *Provisioning*, one for *Reconfiguring* and one for *Retirement*. You can now set these fields to *Embedded Workflows* instead of *Embedded Automate*. You can set the point field to the workflow that you want to use for any, or all, of the three service entry points.

I have added this line to the referenced earlier section: 
> You can set the point field to the workflow that you want to use for any, or all, of the three service entry points.

   ![Entry Points Dynamic](../images/embedworkflow_entrypointsembedded.png){: caption="Entry Points Dynamic" caption-side="bottom"}

4. When you complete all your edits, click **Save**.
5. Confirm that you can order the service catalog item and that it runs the workflows to drive the dynamic dialog dropdown.

<!--
##### Add docker images to a podified deployment and appliances (workflow registries)

> -->

Reviewer: This probably belongs https://github.com/ManageIQ/manageiq-documentation/pull/1752/files#diff-a944ad335565b3b6f0c1776d24ce585d3b42531e2d45bc20a5eaab1950a871e4R281 we'll want to cover building docker containers (and probably point to other e.g. Docker documentation) as well as how to push to a registry and add the registry to the appliance/kubernetes system.

Author: I have moved this comment to the suggested location.
